Deferred/AnyFuture.swift

```diff
@@ -0,0 +1,113 @@
+//
+//  AnyFuture.swift
+//  Deferred
+//
+//  Created by Zachary Waldowski on 8/29/15.
+//  Copyright © 2014-2015 Big Nerd Ranch. Licensed under MIT.
+//
+
+import Dispatch
+
+/*
+The types in this file provide an optimized implementation of type erasure
+for `FutureType`. The techniques were derived from experimenting with
+`AnySequence` and `Mirror` in a playground, and the following post:
+ - https://realm.io/news/type-erased-wrappers-in-swift/
+*/
+
+// Abstract class that fake-conforms to `FutureType` for use by `AnyFuture`.
+class FutureBoxBase<Value>: FutureType {
+    func upon(queue: dispatch_queue_t, body: Value -> ()) {
+        fatalError()
+    }
+
+    func wait(time: Timeout) -> Value? {
+        fatalError()
+    }
+}
+
+// Concrete future wrapper given an instance of a `FutureType`.
+private final class AnyFutureBox<Future: FutureType>: FutureBoxBase<Future.Value> {
+    let base: Future
+    init(base: Future) {
+        self.base = base
+        super.init()
+    }
+
+    override func upon(queue: dispatch_queue_t, body: Future.Value -> ()) {
+        return base.upon(queue, body: body)
+    }
+
+    override func wait(time: Timeout) -> Future.Value? {
+        return base.wait(time)
+    }
+}
+
+// Concrete future wrapper for an always-filled future.
+private final class FilledFutureBox<Value>: FutureBoxBase<Value> {
+    let value: Value
+    init(value: Value) {
+        self.value = value
+        super.init()
+    }
+
+    override func upon(queue: dispatch_queue_t, body: Value -> ()) {
+        dispatch_async(queue) {
+            body(self.value)
+        }
+    }
+
+    override func wait(time: Timeout) -> Value? {
+        return value
+    }
+}
+
+/// A type-erased wrapper over any future.
+///
+/// Forwards operations to an arbitrary underlying future having the same
+/// `Value` type, hiding the specifics of the underlying `FutureType`.
+///
+/// Authors can use this type to:
+///
+/// - Prevent clients from coupling to the specific kind of `FutureType` your
+///   implementation is currently using.
+/// - Publicly expose only the `FutureType` aspect of a deferred value,
+///   ensuring that only your implementation can fill the deferred value
+///   using the `PromiseType` aspect.
+public struct AnyFuture<Value>: FutureType {
```

Comment:
I wonder if  should go ahead and name this `Future`. Yeah, sure, it's out of sync with the Swift standard library, but it's essentially what we want.